pumpkin-protocol/src/client/play/c_commands.rs

```diff
@@ -0,0 +1,382 @@
+        bytebuf.put_list(&self.children, |bytebuf, child| bytebuf.put_var_int(child));
+
+        // redirect node
+        if flags & Self::FLAG_HAS_REDIRECT != 0 {
+            bytebuf.put_var_int(&1.into());
+        }
+
+        // name
+        match self.node_type {
+            ProtoNodeType::Argument { name, .. } | ProtoNodeType::Literal { name, .. } => {
+                bytebuf.put_string_len(name, 32767)
+            }
+            ProtoNodeType::Root => {}
+        }
+
+        // parser id + properties
+        if let ProtoNodeType::Argument {
+            name: _,
+            is_executable: _,
+            parser,
+            override_suggestion_type: _,
+        } = &self.node_type
+        {
+            parser.write_to_buffer(bytebuf)
+        }
+
+        if flags & Self::FLAG_HAS_SUGGESTION_TYPE != 0 {
+            match &self.node_type {
+                ProtoNodeType::Argument { name: _, is_executable: _, parser: _, override_suggestion_type } => {
+                    // suggestion type
+                    let suggestion_type = &override_suggestion_type.expect("ProtoNode::FLAG_HAS_SUGGESTION_TYPE should only be set if override_suggestion_type is not None.");
+                    bytebuf.put_string(suggestion_type.identifier());
+                },
+                _ => unimplemented!("ProtoNode::FLAG_HAS_SUGGESTION_TYPE is only implemented for ProtoNodeType::Argument"),
+            }
+        }
+    }
+}
+
+/// see https://wiki.vg/Command_Data for meaning of flags/enums/etc.
+#[derive(Debug, Clone)]
+pub enum ProtoCmdArgParser<'a> {
+    Bool,
+    Float {
+        min: Option<f32>,
+        max: Option<f32>,
+    },
+    Double {
+        min: Option<f64>,
+        max: Option<f64>,
+    },
+    Integer {
+        min: Option<i32>,
+        max: Option<i32>,
+    },
+    Long {
+        min: Option<i64>,
+        max: Option<i64>,
+    },
+    String(StringProtoArgBehavior),
+    Entity {
+        flags: i8,
+    },
+    GameProfile,
+    BlockPos,
+    ColumnPos,
+    Vec3,
+    Vec2,
+    BlockState,
+    BlockPredicate,
+    ItemStack,
+    ItemPredicate,
+    Color,
+    Component,
+    Style,
+    Message,
+    Nbt,
+    NbtTag,
+    NbtPath,
+    Objective,
+    ObjectiveCriteria,
+    Operation,
+    Particle,
+    Angle,
+    Rotation,
+    ScoreboardSlot,
+    ScoreHolder {
+        flags: i8,
+    },
+    Swizzle,
+    Team,
+    ItemSlot,
+    /// missing on wiki.vg but might be some kind of container slot?
+    ContainerSlot,
+    ResourceLocation,
+    Function,
+    EntityAnchor,
+    IntRange,
+    FloatRange,
+    Dimension,
+    Gamemode,
+    Time {
+        min: i32,
+    },
+    ResourceOrTag {
+        identifier: &'a str,
+    },
+    ResourceOrTagKey {
+        identifier: &'a str,
+    },
+    Resource {
+        identifier: &'a str,
+    },
+    ResourceKey {
+        identifier: &'a str,
+    },
+    TemplateMirror,
+    TemplateRotation,
+    Heightmap,
+    Uuid,
+}
```

Comment:
This Enum is quiet large and can change from time to time trough updates, I think generating and JSON of these Types in our Extractor would be an better soultion

Reply:
Yes, but I'd need you to extract the "command_argument_type" registry for that. This would also mean losing type safety regarding the properties tho.

I'm not sure how you'd want to extract these properties anyways, since the registry just provides objects that implement an interface for reading/writing these. Could you maybe do something with that Template interface?